### PR TITLE
Axe args4j

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Python interface changes:
 - filter_variants_all -> drop_variants
 - filter_samples_all -> drop_samples
 - renamed "condition" to "expr" in parameter names where appropriate

Removed gradle installDist

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 1]
```diff
@@ -0,0 +1,1360 @@
...
+  private def makeSchemaForKudu(vaSignature: Type): StructType =
+    StructType(Array(
+      StructField("variant", Variant.schema, nullable = false),
+      StructField("annotations", vaSignature.schema, nullable = false),
+      StructField("gs", GenotypeStream.schema, nullable = false),
+      StructField("sample_group", StringType, nullable = false)
+    ))
+}
+
+class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) extends AnyVal {
```
Comment (cseed) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: I was thinking of deleting the `VariantDataset` type alias and just make this `VariantDataset` and have the `HailContext` functions.  What's the barrier to doing this now?

File: src/main/scala/is/hail/HailContext.scala [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        s"$k=$v"
+      }.mkString(", ")
+    }")
+
+    val sqlContext = new org.apache.spark.sql.SQLContext(sparkContext)
+    HailContext(sparkContext, sqlContext, tmpDir, branchingFactor)
+  }
+}
+
+case class HailContext private(sc: SparkContext,
```
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: Seems strange for this to be a case class (with a private constructor).

File: python/hail/dataset.py [hunk 3]
```diff
@@ -427,25 +442,27 @@ def annotate_samples_expr(self, condition):
...
         :return: Annotated dataset.
         :rtype: :class:`.VariantDataset`
         """
 
-        if isinstance(condition, list):
-            condition = ','.join(condition)
-        pargs = ['annotatesamples', 'expr', '-c', condition]
-        return self.hc._run_command(self, pargs)
+        if isinstance(expr, list):
+            expr = ','.join(expr)
```
Comment (cseed) [inline python/hail/dataset.py]: You did!  Ah, very nice.  But you have to pluralize.  Looks good.

File: python/hail/dataset.py [hunk 4]
```diff
@@ -493,14 +510,11 @@ def annotate_samples_fam(self, input, quantpheno=False, delimiter='\\\\s+', root
...
         """
 
-        pargs = ['annotatesamples', 'fam', '-i', input, '--root', root, '--missing', missing]
-        if quantpheno:
-            pargs.append('--quantpheno')
-        if delimiter:
-            pargs.append('--delimiter')
-            pargs.append(delimiter)
-        return self.hc._run_command(self, pargs)
+        ffc = env.hail.io.plink.FamFileConfig(quantpheno, delimiter, missing)
```
Comment (tpoterba) [inline python/hail/dataset.py]: Will leave this for a separate PR

File: python/hail/dataset.py [hunk 5]
```diff
@@ -2083,23 +1976,10 @@ def impute_sex(self, maf_threshold=0.0, include_par=False, female_threshold=0.2,
...
-            pargs.append('--female-threshold')
-            pargs.append(str(female_threshold))
-        if male_threshold:
-            pargs.append('--male-threshold')
-            pargs.append(str(male_threshold))
-        if pop_freq:
-            pargs.append('--pop-freq')
-            pargs.append(pop_freq)
-        return self.hc._run_command(self, pargs)
+        jvds = self._jvdf.imputeSex(maf_threshold, include_par, female_threshold, male_threshold, joption(pop_freq))
```
Comment (tpoterba) [inline python/hail/dataset.py]: I assigned where it looked to long to do it without an assignment.  I'd like to leave it that way, if you're OK with that.

File: python/hail/context.py [hunk 6]
```diff
@@ -340,10 +281,12 @@ def import_keytable(self, path, key_names, npartitions=None, config=None):
             config = TextTableConfig()
 
         return KeyTable(self, self._hail.keytable.KeyTable.importTextTable(
-            self._jsc, jarray(self._jvm.java.lang.String, path_args), key_names, npartitions,
+            self._jhc, jarray(self._jvm.java.lang.String, path_args), key_names, npartitions,
```
Comment (tpoterba) [inline python/hail/context.py]: Good idea, did that.

File: python/hail/context.py [hunk 7]
```diff
@@ -264,8 +220,7 @@ def import_gen(self, path, sample_file=None, tolerance=0.02, npartitions=None, c
         :param path: .gen files to import.
         :type path: str or list of str
 
-        :param sample_file: The sample file.
-        :type sample_file: str or None
+        :param str sample_file: The sample file.
```
Comment (tpoterba) [inline python/hail/context.py]: Looks like this was a bug.  It's required on the scala side.

File: python/hail/dataset.py (hunk 4 above)
Comment (cseed) [inline python/hail/dataset.py]: Cool.

File: python/hail/dataset.py (hunk 5 above)
Comment (cseed) [inline python/hail/dataset.py]: Basically I don't care much.  I saw some I would have changed, but the width can be deceptive in the Github diffs.

**Comment by tpoterba** [comment]
Addressed comments.
